src/peft/tuners/vera/layer.py

```diff
@@ -0,0 +1,471 @@
+from typing import List, Optional, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from transformers.pytorch_utils import Conv1D
+
+from peft.tuners.tuners_utils import BaseTunerLayer
+from peft.utils.other import transpose
+
+from .buffer_dict import BufferDict
+
+
+class VeraLayer(BaseTunerLayer):
+    # List all names of layers that may contain adapter weights
+    adapter_layer_names = ("vera_lambda_b", "vera_lambda_d")
+    other_param_names = ("vera_A", "vera_B")
+
+    def __init__(self, base_layer: nn.Module, **kwargs):
+        self.base_layer = base_layer
+        self.r = {}
+        self.vera_dropout = nn.ModuleDict({})
+
+        # For storing vector scale
+        self.vera_lambda_b = nn.ParameterDict({})
+        self.vera_lambda_d = nn.ParameterDict({})
+
+        # Stores a reference to the vera_A/B BufferDict.
+        # Set to `None` otherwise to avoid computation with random weights
+        self.vera_A = None
+        self.vera_B = None
+
+        # Mark the weight as unmerged
+        self._disable_adapters = False
+        self.merged_adapters = []
+
+        base_layer = self.get_base_layer()
+        if isinstance(base_layer, nn.Linear):
+            in_features, out_features = base_layer.in_features, base_layer.out_features
+        elif isinstance(base_layer, nn.Embedding):
+            in_features, out_features = base_layer.num_embeddings, base_layer.embedding_dim
+        elif isinstance(base_layer, Conv1D):
+            in_features, out_features = (
+                base_layer.weight.ds_shape if hasattr(base_layer.weight, "ds_shape") else base_layer.weight.shape
+            )
+
+        self.in_features = in_features
+        self.out_features = out_features
+        self.kwargs = kwargs
+
+    @property
+    def merged(self) -> bool:
+        return bool(self.merged_adapters)
+
+    def update_layer(
+        self,
+        adapter_name,
+        vera_A: BufferDict,
+        vera_B: BufferDict,
+        r,
+        vera_dropout,
+        init_vera_weights,
+        d_initial: float = 1.0,
+    ):
+        if r <= 0:
+            raise ValueError(f"`r` should be a positive integer value but the value passed is {r}")
+        self.r[adapter_name] = r
+        if vera_dropout > 0.0:
+            vera_dropout_layer = nn.Dropout(p=vera_dropout)
+        else:
+            vera_dropout_layer = nn.Identity()
+
+        self.vera_dropout.update(nn.ModuleDict({adapter_name: vera_dropout_layer}))
+        # Actual trainable parameters
+        self.vera_lambda_b[adapter_name] = nn.Parameter(torch.ones(self.out_features), requires_grad=True)
+        self.vera_lambda_d[adapter_name] = nn.Parameter(torch.ones(r), requires_grad=True)
+
+        # non trainable references to vera_A/B buffers
+        # use setattr as this happens post `nn.Module.__init__`
+        # but should not be issue as these are just references to the normally initialised `vera_A/B`
+        setattr(self, "vera_A", vera_A)
```

Comment:
Calling `setattr` here and below should not be necessary, `self.vera_A = vera_A` should be functionally identical, right?